File: examples/cpp/NeuralDepth/neural_depth.cpp
#include <atomic>
#include <csignal>
#include <iostream>
#include <opencv2/opencv.hpp>

#include "depthai/depthai.hpp"

// Global flag for graceful shutdown
std::atomic<bool> quitEvent(false);

void signalHandler(int signum) {
    quitEvent = true;
}

int main() {
    // Set up signal handlers for clean exit
    signal(SIGTERM, signalHandler);
    signal(SIGINT, signalHandler);

    constexpr float FPS = 10.0f;

    // Create pipeline
    dai::Pipeline pipeline;

    // Define sources
    auto cameraLeft = pipeline.create<dai::node::Camera>();
    cameraLeft->build(dai::CameraBoardSocket::CAM_B, std::nullopt, FPS);

    auto cameraRight = pipeline.create<dai::node::Camera>();
    cameraRight->build(dai::CameraBoardSocket::CAM_C, std::nullopt, FPS);

    // Get full resolution outputs
    auto leftOutput = cameraLeft->requestFullResolutionOutput();
    auto rightOutput = cameraRight->requestFullResolutionOutput();

    // Create and build NeuralDepth node
    auto neuralDepth = pipeline.create<dai::node::NeuralDepth>();
    neuralDepth->build(*leftOutput, *rightOutput, dai::DeviceModelZoo::NEURAL_DEPTH_LARGE);

    // Create output queues
    auto confidenceQueue = neuralDepth->confidence.createOutputQueue();
    auto edgeQueue = neuralDepth->edge.createOutputQueue();
    auto disparityQueue = neuralDepth->disparity.createOutputQueue();

    // Create input queue for runtime configuration
    auto inputConfigQueue = neuralDepth->inputConfig.createInputQueue();

    // Start the pipeline
    pipeline.start();

    // Variables for visualization
    double maxDisparity = 1.0;
    cv::Mat colorMap;
    cv::Mat gray(256, 1, CV_8UC1);
    for(int i = 0; i < 256; i++) {
        gray.at<uchar>(i) = i;
    }
    cv::applyColorMap(gray, colorMap, cv::COLORMAP_JET);
    colorMap.at<cv::Vec3b>(0) = cv::Vec3b(0, 0, 0);  // Set zero-disparity pixels to black

    // Get the initial configuration
    auto currentConfig = neuralDepth->initialConfig;

    std::cout << "For adjusting thresholds, use keys:" << std::endl;
    std::cout << " - 'w': Increase confidence threshold" << std::endl;
    std::cout << " - 's': Decrease confidence threshold" << std::endl;
    std::cout << " - 'd': Increase edge threshold" << std::endl;
    std::cout << " - 'a': Decrease edge threshold" << std::endl;
    std::cout << " - 't': Toggle temporal filtering" << std::endl;

    while(!quitEvent && pipeline.isRunning()) {
        // Get confidence data and display it
        auto confidenceData = confidenceQueue->get<dai::ImgFrame>();
        cv::Mat npConfidence = confidenceData->getFrame();
        cv::Mat colorizedConfidence;
        cv::applyColorMap(npConfidence, colorizedConfidence, colorMap);
        cv::imshow("confidence", colorizedConfidence);

        // Get edge data and display it
        auto edgeData = edgeQueue->get<dai::ImgFrame>();
        cv::Mat npEdge = edgeData->getFrame();
        cv::Mat colorizedEdge;
        cv::applyColorMap(npEdge, colorizedEdge, colorMap);
        cv::imshow("edge", colorizedEdge);

        // Get disparity data, normalize, and display it
        auto disparityData = disparityQueue->get<dai::ImgFrame>();
        cv::Mat npDisparity = disparityData->getFrame();

        double minVal, curMax;
        cv::minMaxLoc(npDisparity, &minVal, &curMax);
        maxDisparity = std::max(maxDisparity, curMax);

        cv::Mat normalized;
        npDisparity.convertTo(normalized, CV_8UC1, 255.0 / maxDisparity);

        cv::Mat colorizedDisparity;
        cv::applyColorMap(normalized, colorizedDisparity, colorMap);
        cv::imshow("disparity", colorizedDisparity);

        // Check for keyboard input
        int key = cv::waitKey(1);
        bool configChanged = false;
        if(key == 'q') {
            break;
        } else if(key == 'w') {  // Increase confidence threshold
            uint8_t currentThreshold = currentConfig->getConfidenceThreshold();
            currentConfig->setConfidenceThreshold((currentThreshold + 5) % 255);
            std::cout << "Setting confidence threshold to: " << (int)currentConfig->getConfidenceThreshold() << std::endl;
            configChanged = true;
        } else if(key == 's') {  // Decrease confidence threshold
            uint8_t currentThreshold = currentConfig->getConfidenceThreshold();
            currentConfig->setConfidenceThreshold(currentThreshold - 5);  // uint8_t will wrap around on underflow
            std::cout << "Setting confidence threshold to: " << (int)currentConfig->getConfidenceThreshold() << std::endl;
            configChanged = true;
        } else if(key == 'd') {  // Increase edge threshold
            uint8_t currentThreshold = currentConfig->getEdgeThreshold();
            currentConfig->setEdgeThreshold((currentThreshold + 1) % 255);
            std::cout << "Setting edge threshold to: " << (int)currentConfig->getEdgeThreshold() << std::endl;
            configChanged = true;
        } else if(key == 'a') {  // Decrease edge threshold
            uint8_t currentThreshold = currentConfig->getEdgeThreshold();
            currentConfig->setEdgeThreshold(currentThreshold - 1);  // uint8_t will wrap around on underflow
            std::cout << "Setting edge threshold to: " << (int)currentConfig->getEdgeThreshold() << std::endl;
            configChanged = true;
        } else if(key == 't') {
            bool enable = !currentConfig->postProcessing.temporalFilter.enable;
            currentConfig->postProcessing.temporalFilter.enable = enable;
            std::cout << "Temporal filtering: " << (enable ? "on" : "off") << std::endl;
            configChanged = true;
        }

        // Send the updated configuration to the device
        if(configChanged) {
            inputConfigQueue->send(currentConfig);
        }
    }

    // Pipeline is stopped automatically when the 'pipeline' object goes out of scope.
    return 0;
}

File: examples/python/NeuralDepth/neural_depth.py
#!/usr/bin/env python3

import cv2
import depthai as dai
import numpy as np

FPS = 25
# Create pipeline
with dai.Pipeline() as pipeline:
    cameraLeft = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_B, sensorFps=FPS)
    cameraRight = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_C, sensorFps=FPS)
    leftOutput = cameraLeft.requestFullResolutionOutput()
    rightOutput = cameraRight.requestFullResolutionOutput()

    neuralDepth = pipeline.create(dai.node.NeuralDepth).build(leftOutput, rightOutput, dai.DeviceModelZoo.NEURAL_DEPTH_LARGE)
    neuralDepth.inputConfig.setBlocking(False)

    confidenceQueue = neuralDepth.confidence.createOutputQueue()
    edgeQueue = neuralDepth.edge.createOutputQueue()
    disparityQueue = neuralDepth.disparity.createOutputQueue()
    depthQueue = neuralDepth.depth.createOutputQueue()

    inputConfigQueue = neuralDepth.inputConfig.createInputQueue()
    currentConfig = neuralDepth.initialConfig
    currentConfig.postProcessing.thresholdFilter.maxRange = 10000

    # Connect to device and start pipeline
    pipeline.start()
    maxDisparity = 1
    colorMap = cv2.applyColorMap(np.arange(256, dtype=np.uint8), cv2.COLORMAP_JET)
    colorMap[0] = [0, 0, 0]  # to make zero-disparity pixels black
    print("For adjusting thresholds, use keys:")
    print(" - 'w': Increase confidence threshold")
    print(" - 's': Decrease confidence threshold")
    print(" - 'd': Increase edge threshold")
    print(" - 'a': Decrease edge threshold")
    print(" - 'i': Increase threshold filter max range")
    print(" - 'k': Decrease threshold filter max range")
    print(" - 'l': Increase threshold filter min range")
    print(" - 'j': Decrease threshold filter min range")
    print(" - 't': Toggle temporal filtering")
    while pipeline.isRunning():
        confidenceData = confidenceQueue.get()
        assert isinstance(confidenceData, dai.ImgFrame)
        npConfidence = confidenceData.getFrame()
        colorizedConfidence = cv2.applyColorMap(((npConfidence)).astype(np.uint8), colorMap)
        cv2.imshow("confidence", colorizedConfidence)

        edgeData = edgeQueue.get()
        assert isinstance(edgeData, dai.ImgFrame)
        npEdge = edgeData.getFrame()
        colorizedEdge = cv2.applyColorMap(((npEdge)).astype(np.uint8), colorMap)
        cv2.imshow("edge", colorizedEdge)


        disparityData = disparityQueue.get()
        assert isinstance(disparityData, dai.ImgFrame)
        npDisparity = disparityData.getFrame()
        maxDisparity = max(maxDisparity, np.max(npDisparity))
        colorizedDisparity = cv2.applyColorMap(((npDisparity / maxDisparity) * 255).astype(np.uint8), colorMap)
        cv2.imshow("disparity", colorizedDisparity)

        depthData = depthQueue.get()
        assert isinstance(depthData, dai.ImgFrame)
        npDepth = depthData.getFrame()
        maxRange = max(currentConfig.postProcessing.thresholdFilter.maxRange, 1)
        depthFrame = np.clip((npDepth / maxRange) * 255, 0, 255).astype(np.uint8)
        colorizedDepth = cv2.applyColorMap(depthFrame, colorMap)
        cv2.imshow("depth", colorizedDepth)

        key = cv2.waitKey(1)
        if key == ord('q'):
            pipeline.stop()
            break
        if key == ord('w'):
            currentThreshold = currentConfig.getConfidenceThreshold()
            currentConfig.setConfidenceThreshold((currentThreshold + 5) % 255)
            print("Setting confidence threshold to:", currentConfig.getConfidenceThreshold())
            inputConfigQueue.send(currentConfig)
        if key == ord('s'):
            currentThreshold = currentConfig.getConfidenceThreshold()
            currentConfig.setConfidenceThreshold((currentThreshold - 5) % 255)
            print("Setting confidence threshold to:", currentConfig.getConfidenceThreshold())
            inputConfigQueue.send(currentConfig)
        if key == ord('d'):
            currentThreshold = currentConfig.getEdgeThreshold()
            currentConfig.setEdgeThreshold((currentThreshold + 1) % 255)
            print("Setting edge threshold to:", currentConfig.getEdgeThreshold())
            inputConfigQueue.send(currentConfig)
        if key == ord('a'):
            currentThreshold = currentConfig.getEdgeThreshold()
            currentConfig.setEdgeThreshold((currentThreshold - 1) % 255)
            print("Setting edge threshold to:", currentConfig.getEdgeThreshold())
            inputConfigQueue.send(currentConfig)
        if key == ord('t'):
            currentConfig.postProcessing.temporalFilter.enable = not currentConfig.postProcessing.temporalFilter.enable
            print("Temporal filtering:", "on" if currentConfig.postProcessing.temporalFilter.enable else "off")
            inputConfigQueue.send(currentConfig)
        if key == ord('i'):
            currentMaxRange = currentConfig.postProcessing.thresholdFilter.maxRange
            currentConfig.postProcessing.thresholdFilter.maxRange = min(currentMaxRange + 100, 65535)
            print("Setting threshold filter max range to:", currentConfig.postProcessing.thresholdFilter.maxRange)
            inputConfigQueue.send(currentConfig)
        if key == ord('k'):
            currentMaxRange = currentConfig.postProcessing.thresholdFilter.maxRange
            minRange = currentConfig.postProcessing.thresholdFilter.minRange
            currentConfig.postProcessing.thresholdFilter.maxRange = max(currentMaxRange - 100, minRange)
            print("Setting threshold filter max range to:", currentConfig.postProcessing.thresholdFilter.maxRange)
            inputConfigQueue.send(currentConfig)
        if key == ord('l'):
            currentMinRange = currentConfig.postProcessing.thresholdFilter.minRange
            maxRange = currentConfig.postProcessing.thresholdFilter.maxRange
            currentConfig.postProcessing.thresholdFilter.minRange = min(currentMinRange + 100, maxRange)
            print("Setting threshold filter min range to:", currentConfig.postProcessing.thresholdFilter.minRange)
            inputConfigQueue.send(currentConfig)
        if key == ord('j'):
            currentMinRange = currentConfig.postProcessing.thresholdFilter.minRange
            currentConfig.postProcessing.thresholdFilter.minRange = max(currentMinRange - 100, 0)
            print("Setting threshold filter min range to:", currentConfig.postProcessing.thresholdFilter.minRange)
            inputConfigQueue.send(currentConfig)
